engine/src/test/java/org/hibernate/validator/test/cfg/ProgrammaticConstraintDefinitionsTest.java

```diff
@@ -61,6 +63,19 @@ public void safeHtmlProgrammaticDefinition() {
 		doProgrammaticTest( new SafeHtmlDef().whitelistType( SafeHtml.WhiteListType.NONE ), "test", 0 );
 		doProgrammaticTest( new SafeHtmlDef().whitelistType( SafeHtml.WhiteListType.RELAXED ), "<td>1234qwer</td>", 0 );
 		doProgrammaticTest( new SafeHtmlDef().whitelistType( SafeHtml.WhiteListType.NONE ).additionalTags( "td" ), "<td>1234qwer</td>", 0 );
+
+		AnnotationDescriptor<SafeHtml.Tag> tagDescriptor = new AnnotationDescriptor( SafeHtml.Tag.class );
+		tagDescriptor.setValue( "name", "td" );
+		tagDescriptor.setValue( "attributes", new String[]{ "class", "id" } );
+		doProgrammaticTest( new SafeHtmlDef().whitelistType( SafeHtml.WhiteListType.NONE )
+				.additionalTagsWithAttributes( AnnotationFactory.create( tagDescriptor ) ), "<td class='class' id='tableId'>1234qwer</td>", 0 );
+
+		AnnotationDescriptor<SafeHtml.Tag> protocolDescriptor = new AnnotationDescriptor( SafeHtml.Tag.class );
+		protocolDescriptor.setValue( "name", "img" );
+		protocolDescriptor.setValue( "attributes", new String[]{ "src" } );
+		protocolDescriptor.setValue( "protocols", new String[]{ "data" } );
+		doProgrammaticTest( new SafeHtmlDef().whitelistType( SafeHtml.WhiteListType.NONE )
```

Comment:
Let's think about it from a user's perspective. That's what I'd like to be able to write:

    new SafeHtmlDef().whitelistType( SafeHtml.WhiteListType.NONE )
        // let me add a tag and configure i all the way
        .additionalTag( "img" )
            .attribute( "src" )
                .protocols( "data", "snafu" )
            .attribute( "ref" )
                .protocols( "bla" )
        .additionalTag( "a" )
            .attribute( "href" )
                .protocols( "data" )
            .attribute( "bla" )
            .attribute( "blub" )
        // fast path for adding tags without further config
        .additionalTags( "foo", "bar" )